### PR TITLE
PaymentFlow Analytics Audit

### PR DESCRIPTION
**Pull request by KunJeongPark**
Summary of changes
These changes are going into the fpti-feature branch which will get merged once we audit all modules and make the switch to the FPTI endpoints

Update PaymentFlow analytic events to prepare switch to FPT
Update tests
Bucketed all analytic events sent before callbacks into 3 categories of final events: paymentFailed, paymentSucceeded, paymentCanceled.
Checklist
- [x] Added a changelog entry

Authors
@KunJeongPark

### CODE REVIEW

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (KunJeongPark) [inline Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:226]: I didn't understand why there was need to do this when apiClient reference is passed into this function. To show separation of concern?

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (KunJeongPark) [inline Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:52]: @scannillo I realize that I am sending "three-d-secure:start-payment:started" here for 3Ds flow which deviates from "3ds:verify:started" that you specified in the list of events for 3DS.

Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:52]: I guess I can have two sets of analytic event names in BTFlowAnalytics and use a helper function to send the appropriate name depending not he flowName.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (scannillo) [inline Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:161]: This result of this boolean value should only be used for analytics, and should not call the merchant's completion ([see PPCP](https://github.com/paypal/iOS-SDK/blob/3be6c44b1b5fad08d219886cf2d3bfeb7435cf40/Sources/PayPalWebPayments/PayPalWebCheckoutClient.swift#L60-L66)).

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (scannillo) [inline Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:170]: How come this isn't using the helper method defined on line 265?

Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:170]: I just wanted to clarify that this is using apiClient parameter in the `handle` function as opposed to the calls in `handleOpen` function where apiClient is not passed in as a parameter, it is referenced as `paymentFlowClientDelegate?.apiClient()`. They are the same but I wanted to keep the same logic.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:161]: We followed the same pattern for PayPalClient. https://github.com/braintree/braintree_ios/pull/977/files#diff-7593900bd413ee81d48fae61aeeab899a00cbfdc0d2a688e21d3acf3f7ef3360R358 Isn't failure to launch this authentication page an end event?

File: UnitTests/BraintreePaymentFlowTests/BTLocalPayment_Tests.swift
Comment (KunJeongPark) [inline UnitTests/BraintreePaymentFlowTests/BTLocalPayment_Tests.swift:295]: I have to figure out a way that this is toggled during the ASWebSession due to system notification.

Comment (KunJeongPark) [reply UnitTests/BraintreePaymentFlowTests/BTLocalPayment_Tests.swift:295]: This test was just to toss an idea but this test will always pass since I am resetting returnedToAppAfterPermit to false before the webSession launches in the code.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (scannillo) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift]: It's fine if the 3DS events are wonky until you get to it in your 3DS PR!

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (scannillo) [inline Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:275]: Nitpick - I personally don't think we need an extension here

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (scannillo) [inline Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:116]:
```suggestion
                        // User canceled by breaking out of the LocalPayment browser switch flow
```

File: UnitTests/BraintreeThreeDSecureTests/BTThreeDSecure_Tests.swift
Comment (KunJeongPark) [reply UnitTests/BraintreeThreeDSecureTests/BTThreeDSecure_Tests.swift:395]: Good idea. I was thinking about order of doing this. Wait for 3DS analytics to get merged?

File: UnitTests/BraintreePaymentFlowTests/BTLocalPayment_Tests.swift
Comment (scannillo) [inline UnitTests/BraintreePaymentFlowTests/BTLocalPayment_Tests.swift:244]: This test can also assert that the `start` event was called

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift]: I've refactored this to return sessionDidDisplay result on successful launch. So we wouldn't be able to send this specific error at that point since we are not returning to merchant app here. I will delete addition of webSessionFailedToLaunch and keep BTPaymentFlowError.webSessionError(error)

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (KunJeongPark) [inline Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:109]: If it's really important to be able to send `webSessionFailedToLaunch` error back to merchant at end of the error block, we can set boolean flag here and reset before calling the webSession.

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (jaxdesmarais) [inline Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:67]: We can rely on an implicit return here and remove the `return` keyword. This has been our preference throughout the SDK.

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (jaxdesmarais) [inline Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:66]: nit: can we group these properties together? It looks like in other files we would do vars > computed vars > weak vars.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (jaxdesmarais) [inline Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:23]: This looks to be internal, is there a reason it's under the `Public Methods` mark?

File: Sources/BraintreePaymentFlow/WebAuthenticationSession.swift
Comment (jaxdesmarais) [inline Sources/BraintreePaymentFlow/WebAuthenticationSession.swift:8]: This is a punt for sure, but I still think it makes sense to move this class into Core to prevent needing to make updates across several modules if we want to make changes to our `WebAuthenticationSession` since they all use the exact same logic. We could potentially be in a scenario where we make an update here but forget to in SEPA and have different behavior across modules that's unintended. Nothing to worry about for this particular PR though. 😄

File: UnitTests/BraintreePaymentFlowTests/BTLocalPayment_Tests.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreePaymentFlowTests/BTLocalPayment_Tests.swift:6]: nit: move this above the `@testable` imports with the other standard imports

File: UnitTests/BraintreePaymentFlowTests/MockWebAuthenticationSession.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreePaymentFlowTests/MockWebAuthenticationSession.swift:5]: Similar comment here on punting on duplication - we can likely move `MockWebAuthenticationSession` into `BraintreeTestShared` and share logic to prevent duplication errors.

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:66]: Not sure what you mean by group the properties?

File: Sources/BraintreePaymentFlow/WebAuthenticationSession.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/WebAuthenticationSession.swift:8]: I did think about that doing SEPA but I think we don't have the two separate completion handler as we do in PPCP.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:23]: you are right, we only need it as public if we move WebAuthenticationSession to Core module, I think,

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (jaxdesmarais) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift]: Right now we have:
```swift
var paymentID: String?
weak var paymentFlowClientDelegate: BTPaymentFlowClientDelegate?
var correlationID: String?
var paymentFlowAPIClient: BTAPIClient? { ... }
```

Should be:
```swift
var paymentID: String?
var correlationID: String?

var paymentFlowAPIClient: BTAPIClient? { ... }

weak var paymentFlowClientDelegate: BTPaymentFlowClientDelegate?
```

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (scannillo) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift]: We should style-guide that about implicit `return`

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift]: Done

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift]: It's done here.

File: UnitTests/BraintreePaymentFlowTests/MockWebAuthenticationSession.swift
Comment (scannillo) [reply UnitTests/BraintreePaymentFlowTests/MockWebAuthenticationSession.swift:5]: Do we have a JIRA for that work?

Comment (KunJeongPark) [reply UnitTests/BraintreePaymentFlowTests/MockWebAuthenticationSession.swift:5]: Not yet. I will make it and link here.

File: Sources/BraintreePaymentFlow/WebAuthenticationSession.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/WebAuthenticationSession.swift:8]: Here is the ticket I made for it. https://paypal.atlassian.net/browse/DTBTSDK-2688

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (jaxdesmarais) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift]: > We should style-guide that about implicit `return`

Added a note about implicit return to the [style guide](https://hackmd.io/kIyOnIFIRrKY4b5wQ0dUpw). Also added a section on use of `self` since we've talked about it here and there but not actually wrote it down.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (jaxdesmarais) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: I don't see the error `webSessionFailedToLaunch` anywhere in this codebase currently aside from in the CHANGELOG and error enum. Would you mind clarifying where this error is being sent?

Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: I see it  in our feature branch I think, on BTPayPalError.

Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: I did have it in previous commit but @scannillo thought that we shouldn't return to merchant app at this point with an error but it should be handled in completion handler. In PayPal module, we did return with completion with error on failure to launch and sent that webSessionFailedToLaunch. So I was wondering what to do with that error. How we make it consistent. Sorry, it took some digging to last week me.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (sshropshire) [inline Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:39]: This should be under `// MARK: - Internal Properties` right?

**Review by sshropshire — APPROVED** [review]

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (jaxdesmarais) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: In PayPal it looks like we send `webSessionFailedToLaunch` if `isAuthenticationSessionStarted = false`. It looks like we don't use `isAuthenticationSessionStarted` in `PaymentFlow` at all so I don't know that we need to add it.

I think for now we can go with what we have but should have a follow up PR to align how we are using `WebAuthenticationSession` as a whole. We shouldn't need 3 different implementations of `WebAuthenticationSession`. I think this brings us back to an earlier discussion we had around having a shared `WebAuthenticationSession` in Core that we use across all modules since we have deviated and have 3 different implementations in SEPA vs PayPal vs PaymentFlow. Even if we don't move it into Core we likely can/should use the same implementation across the modules unless there is a compelling reason not to.

File: CHANGELOG.md
Comment (scannillo) [inline CHANGELOG.md:33]: This CHANGELOG entry needs to go under `unreleased`. Here, it was added under the 6.0.0-beta3 release, which already went out. The changes you're making in this PR apply to the next unreleased version, which will be tagged upon release.

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (scannillo) [inline Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:62]: Nitpick
```suggestion
    var apiClient: BTAPIClient? {
```

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (scannillo) [inline Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:177]: This event is missing from our event spreadsheet. Can you add it?

Also, it is missing from SEPA & 3DS as well.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (scannillo) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: ➕ 1

Comment (scannillo) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: Victoria - we don't want to send a tokenization fail event for the bool status of the ASWeb. That is because we are not calling the merchant's completion in that case. I don't think that bool value is reliable enough.

There is a discrepancy with the PayPal implementation [here](https://github.com/braintree/braintree_ios/blob/9cb3aa0b50b027e63f93a8abc4cda6492d897b1a/Sources/BraintreePayPal/BTPayPalClient.swift#L357). This PayPal code linked is incorrect and needs to be updated.

Not calling a merchant completion nor sending a tokenize failed event for the `asWeb.start()` bool result matches our [v5 implementation](https://github.com/braintree/braintree_ios/blob/8f34d3f1deada3559ad287d41d0fb7c65de2c9b3/Sources/BraintreePayPal/BTPayPalDriver.m#L299-L305).

Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: The reason I made separate implementation here was that the PayPal implementation did not have separate completion for` sessionDidDisplay` and `sessionDidComplete`. At the time I did implementation for PayPal, I was not aware that we were conforming to PPCP way of doing this.

 I didn't think it would matter since if not failed here with merchant completion here, it will be caught in the completion as it is with the implementation without sending merchant completion fail here. I do agree that we want consistency but I think either way, failure to display returns to merchant app with fail eventually.

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:177]: So we do want this for all of the modules? I was following the spreadsheet for other modules.

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: Here are questions I have remaining: 
1> Do we want to send networkConnectionLost analytics for all the modules? Sammy, I think for one of the modules, we decide not to and I had a lapse for LocalPayment module and included it.
2> Should I make a separate ticket for not returning to merchant completion right away with sessionDidDisplay false in PayPal module.?
3> Do we make a separate ticket for unified websession module?
Do we want to include refactoring PayPal module to use this wrapper class for ASWebSession instead of using ASWebSession directly?

Comment (jaxdesmarais) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: > 1> Do we want to send networkConnectionLost analytics for all the modules? Sammy, I think for one of the modules, we decide not to and I had a lapse for LocalPayment module and included it.

I think this is up to us to decide. The merchant that requested this to be added only uses PayPal and Venmo through us. It may make sense to add it to the missing modules in a follow up PR as it looks like we have it in about half the modules now.

> 2> Should I make a separate ticket for not returning to merchant completion right away with sessionDidDisplay false in PayPal module?

We should make a follow up PR as a whole to align our usage of ASWebAuthenticationSession as a whole. All of our modules depend on Core so I think it makes sense to move the implementation of our WebAuthenticationSession into there in a follow up PR.

> 3> Do we make a separate ticket for unified websession module?
Do we want to include refactoring PayPal module to use this wrapper class for ASWebSession instead of using ASWebSession directly?

Yes, this should be in a follow up PR and not in this PR.

Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: Thank you Jax. So should those be two separate PR's - one for not returning to merchant fail completion after failure to display in PayPal module, and one for refactoring PayPal modulate make unified WebSession class?

Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: And networkConnectionLost, if we decide to include, we should have another PR to add them all in.

Comment (jaxdesmarais) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: > Thank you Jax. So should those be two separate PR's - one for not returning to merchant fail completion after failure to display in PayPal module, and one for refactoring PayPal module to use wrapper class for web session and make unified WebSession class for all three modules?

Yep, we should have a follow up PR to align our usage and implementation of ASWebAuthenticationSession.

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (scannillo) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:177]: I updated the spreadsheet to include networkLost events for all modules!

File: Sources/BraintreePaymentFlow/BTPaymentFlowError.swift
Comment (scannillo) [inline Sources/BraintreePaymentFlow/BTPaymentFlowError.swift:91]:
```suggestion
               return "ASWebAuthenticationSession failed with \(error.localizedDescription)"
```

File: UnitTests/BraintreePaymentFlowTests/BTPaymentFlowAnalytics_Tests.swift
Comment (scannillo) [inline UnitTests/BraintreePaymentFlowTests/BTPaymentFlowAnalytics_Tests.swift:19]:
```suggestion
```

**Review by scannillo — APPROVED** [review]

File: UnitTests/BraintreeThreeDSecureTests/BTThreeDSecureClient_Tests.swift
Comment (jaxdesmarais) [inline UnitTests/BraintreeThreeDSecureTests/BTThreeDSecureClient_Tests.swift:473]: We should make sure to address these TODOs and commented out tests

File: Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift
Comment (KunJeongPark) [reply Sources/BraintreePaymentFlow/BTLocalPaymentRequest.swift:177]: Here is the Jira ticket for adding networkConnectionLost event for all modules: https://paypal.atlassian.net/browse/DTBTSDK-2711

File: Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift
Comment (jaxdesmarais) [reply Sources/BraintreePaymentFlow/BTPaymentFlowClient.swift:107]: Resolving: ticket for unified ASWeb logic is DTBTSDK-2712

File: UnitTests/BraintreeThreeDSecureTests/BTThreeDSecureClient_Tests.swift
Comment (KunJeongPark) [reply UnitTests/BraintreeThreeDSecureTests/BTThreeDSecureClient_Tests.swift:473]: I will take care of this in the 3DS analytic PR after merging in the updated fpti-feature there.

**Pull request merged**
